doc/Programming-guide/Using-Kaa-endpoint-SDKs/C/SDK-ESP8266/attach/esp8266-sample/driver/uart.c

```diff
@@ -0,0 +1,283 @@
+/*
+ *  Copyright (C) 2014 -2016  Espressif System
+ *
+ */
+
+#include "esp_common.h"
+
+#include "freertos/FreeRTOS.h"
+#include "freertos/task.h"
+#include "freertos/queue.h"
+
+#include "uart.h"
+
+enum {
+    UART_EVENT_RX_CHAR,
+    UART_EVENT_MAX
+};
+
+typedef struct _os_event_ {
+    uint32 event;
+    uint32 param;
+} os_event_t;
+
+xTaskHandle xUartTaskHandle;
+xQueueHandle xQueueUart;
+
+LOCAL STATUS
+uart_tx_one_char(uint8 uart, uint8 TxChar)
+{
+    while (true) {
+        uint32 fifo_cnt = READ_PERI_REG(UART_STATUS(uart)) & (UART_TXFIFO_CNT << UART_TXFIFO_CNT_S);
+
+        if ((fifo_cnt >> UART_TXFIFO_CNT_S & UART_TXFIFO_CNT) < 126) {
+            break;
+        }
+    }
+
+    WRITE_PERI_REG(UART_FIFO(uart), TxChar);
+    return OK;
+}
+
+LOCAL void
+uart1_write_char(char c)
+{
+    if (c == '\n') {
+        uart_tx_one_char(UART1, '\r');
+        uart_tx_one_char(UART1, '\n');
+    } else if (c != '\r') {
+        uart_tx_one_char(UART1, c);
+    }
+}
+
+void
+uart0_write_char(char c)
+{
+    if (c == '\n') {
+        uart_tx_one_char(UART0, '\r');
+        uart_tx_one_char(UART0, '\n');
+    } else if (c != '\r') {
+        uart_tx_one_char(UART0, c);
+    }
+}
```

Comment:
```c
void uart_write_char(uart, char c)
{
    if (c == '\n') {
        uart_tx_one_char(uart, '\r');
        uart_tx_one_char(uart, '\n');
    } else if (c != '\r') {
        uart_tx_one_char(uart, c);
    }
}

void uart0_write_char(char c)
{
    uart_write_char(UART0, c);
}

void uart1_write_char(char c)
{
    uart_write_char(UART1, c);
}
```

It may be also possible `uart0/1_write_char` functions won't be needed.